File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
class Solution {
    public int solution(String dartResult) {
        char[] results = dartResult.toCharArray();
        Stack<Integer> stack = new Stack<>();
        for (int i=0; i<results.length; i++) {
            if (results[i] == 'S') {
                int num = stack.pop();
                stack.push((int)Math.pow(num, 1));
            } else if (results[i] == 'D') {
                int num = stack.pop();
                stack.push((int)Math.pow(num, 2));
            } else if (results[i] == 'T') {
                int num = stack.pop();
                stack.push((int)Math.pow(num, 3));
            } else if (results[i] == '*') {
                if (stack.size() > 1) {
                    int first = stack.pop() * 2;
                    int second = stack.pop() * 2;
                    stack.push(second);
                    stack.push(first);
                } else {
                    int first = stack.pop() * 2;
                    stack.push(first);
                }
            } else if (results[i] == '#') {
                stack.push(stack.pop() * -1);
            } else {
                if (i > 0 && Character.isDigit(results[i-1])) {
                    stack.pop();
                    stack.push(10);
                } else {
                    stack.push(results[i] - '0');  
                }
            }
        }
        int sum = 0;
        for (int i : stack) {
            sum += i;
        }
        return sum;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    let answer = 0;
    let record = [];
    const bonus = {
        'S': 1,
        'D': 2,
        'T': 3
    }
    
    let dr = dartResult.split(/([SDT*#])/g);
    for (d of dr) {
        if (d === '') continue;
        let len = record.length;
        
        if (d === 'S' || d === 'D' || d === 'T') { // 보너스
            record[len-1] = Math.pow(record[len-1], bonus[d]);
        } else if (d === '*' || d === '#') { // 옵션
            if (d === '*') {
                if (len === 0) continue;
                record[len-1] *= 2;
                if (len < 2) continue;
                record[len-2] *= 2;
            } else if (d === '#') {
                if (len === 0) continue;
                record[len-1] = -record[len-1];
            }
        } else { // 점수
            record.push(Number(d));
        }
    }
    
    for (r of record) {
        answer += r;
    }
    
    return answer;
}